src/realm/impl/copy_replication.cpp

```diff
@@ -0,0 +1,276 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#include "copy_replication.hpp"
+#include <realm/list.hpp>
+#include <realm/set.hpp>
+#include <realm/dictionary.hpp>
+
+namespace realm::impl {
+
+void CopyReplication::add_class(TableKey, StringData name, bool is_embedded)
+{
+    if (auto existing_table = m_tr->get_table(name)) {
+        if (existing_table->is_embedded() != is_embedded)
+            throw std::runtime_error(util::format("Incompatible class: %1", name));
+        return;
+    }
+    if (is_embedded) {
+        m_tr->add_embedded_table(name);
+    }
+    else {
+        m_tr->add_table(name);
+    }
+}
+
+void CopyReplication::add_class_with_primary_key(TableKey, StringData name, DataType type, StringData pk_name,
+                                                 bool nullable)
+{
+    if (auto existing_table = m_tr->get_table(name)) {
+        auto pk_col = existing_table->get_primary_key_column();
+        if (DataType(pk_col.get_type()) != type || existing_table->get_column_name(pk_col) != pk_name)
+            throw std::runtime_error(util::format("Incompatible class: %1", name));
+        return;
+    }
+    m_tr->add_table_with_primary_key(name, type, pk_name, nullable);
+}
+
+void CopyReplication::insert_column(const Table* t, ColKey col_key, DataType type, StringData name, Table* dest)
+{
+    m_current.table = t;
+    auto table = get_table_in_destination_realm();
+    if (ColKey existing_key = table->get_column_key(name)) {
+        if (existing_key.get_type() != col_key.get_type() || existing_key.get_attrs() != col_key.get_attrs())
+            throw std::runtime_error(util::format("Incompatible property: %1::%2", t->get_name(), name));
+        return;
+    }
+    if (dest) {
+        auto target_table_name = dest->get_name();
+        auto target = m_tr->get_table(target_table_name);
+        if (col_key.is_list()) {
+            table->add_column_list(*target, name);
+        }
+        else if (col_key.is_set()) {
+            table->add_column_set(*target, name);
+        }
+        else if (col_key.is_dictionary()) {
+            table->add_column_dictionary(*target, name);
+        }
+        else {
+            table->add_column(*target, name);
+        }
+    }
+    else {
+        if (col_key.is_list()) {
+            table->add_column_list(type, name, col_key.is_nullable());
+        }
+        else if (col_key.is_set()) {
+            table->add_column_set(type, name, col_key.is_nullable());
+        }
+        else if (col_key.is_dictionary()) {
+            auto key_type = t->get_dictionary_key_type(col_key);
+            table->add_column_dictionary(type, name, col_key.is_nullable(), key_type);
+        }
+        else {
+            table->add_column(type, name, col_key.is_nullable());
```

Comment:
can we add an index to the new column if it exists in the source table?

Reply:
Sure!